File: Two_Pointer/438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        pn = len(p)
        sn = len(s)
        if pn > sn:
            return []
        
        pcount = {}
        scount = {}
        for i in range(pn):
            pcount[p[i]] = 1 + pcount.get(p[i], 0)
            scount[s[i]] = 1 + scount.get(s[i], 0)

        ans = []
        j = 0
        for i in range(pn, sn):
            if pcount == scount:
                ans.append(j)
            
            scount[s[i]] = 1 + scount.get(s[i], 0)
            scount[s[j]] -= 1

            if scount[s[j]] == 0:
                scount.pop(s[j])
            j += 1
        if pcount == scount:
                ans.append(j)
        
        return ans

File: Two_Pointer/438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int pn = p.length();
        int sn = s.length();
        
        if (pn > sn) {
            return {};
        }
        
        unordered_map<char, int> pcount, scount;
        for (int i = 0; i < pn; ++i) {
            pcount[p[i]]++;
            scount[s[i]]++;
        }

        vector<int> ans;
        int j = 0;
        
        for (int i = pn; i < sn; ++i) {
            if (pcount == scount) {
                ans.push_back(j);
            }
            
            scount[s[i]]++;
            scount[s[j]]--;
            
            if (scount[s[j]] == 0) {
                scount.erase(s[j]);
            }
            j++;
        }
        
        if (pcount == scount) {
            ans.push_back(j);
        }
        
        return ans;
    }
};